etcdserver/api/v3rpc/watch.go

```diff
@@ -61,12 +61,24 @@ func (ws *watchServer) Watch(stream pb.Watch_WatchServer) error {
 func sendLoop(stream pb.Watch_WatchServer, watcher storage.Watcher, closec chan struct{}) {
 	for {
 		select {
-		case e := <-watcher.Chan():
+		case e, ok := <-watcher.Chan():
+			if !ok {
+				return
+			}
 			err := stream.Send(&pb.WatchResponse{Event: &e})
+			storage.ReportEventReceived()
```

Comment:
I do not think so. Received means "delivered" to the other side. Probably the func needs a better naming. 

We will add failed event soon. It will replace the report a few lines below.